File: trees/invert-binary-tree.js
// https://leetcode.com/problems/invert-binary-tree/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {TreeNode}
     */
    invertTree(root) {
        if (root === null) return null;  // Base case: empty tree

        // Swap left and right children
        const temp = root.left;
        root.left = root.right;
        root.right = temp;

        // Recurse for left and right subtrees
        this.invertTree(root.left);
        this.invertTree(root.right);

        return root;
    }
}

File: python/trees/invert_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# start off with the root
# we want to end with the root but inverted
# base case and recursive case
# implement BFS, traverse the right side, then the left side


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # base case
        if not root:
            return None

        # recursive case
        root.left, root.right = root.right, root.left

        self.invertTree(root.right)
        self.invertTree(root.left)

        return root